h2/src/test/org/h2/test/jdbc/TestCustomDataTypesHandler.java

```diff
@@ -0,0 +1,511 @@
+            return new ValueComplex(val);
+        }
+
+        /** {@inheritDoc} */
+        @Override
+        public String getSQL() {
+            return val.toString();
+        }
+
+        /** {@inheritDoc} */
+        @Override
+        public int getType() {
+            return TestOnlyCustomDataTypesHandler.COMPLEX_DATA_TYPE_ID;
+        }
+
+        /** {@inheritDoc} */
+        @Override
+        public long getPrecision() {
+            return 0;
+        }
+
+        /** {@inheritDoc} */
+        @Override
+        public int getDisplaySize() {
+            return 0;
+        }
+
+        /** {@inheritDoc} */
+        @Override
+        public String getString() {
+            return val.toString();
+        }
+
+        /** {@inheritDoc} */
+        @Override
+        public Object getObject() {
+            return val;
+        }
+
+        /** {@inheritDoc} */
+        @Override
+        public void set(PreparedStatement prep, int parameterIndex) throws SQLException {
+            Object obj = JdbcUtils.deserialize(getBytesNoCopy(), getDataHandler());
+            prep.setObject(parameterIndex, obj, Types.JAVA_OBJECT);
+        }
+
+        /** {@inheritDoc} */
+        @Override
+        protected int compareSecure(Value v, CompareMode mode) {
+            return val.compare((ComplexNumber) v.getObject());
+        }
+
+        /** {@inheritDoc} */
+        @Override
+        public int hashCode() {
+            return val.hashCode();
+        }
+
+        /** {@inheritDoc} */
+        @Override
+        public boolean equals(Object other) {
+            if (other == null) {
+                return false;
+            }
+            if (!(other instanceof ValueComplex)) {
+                return false;
+            }
+            ValueComplex complex = (ValueComplex)other;
+            return complex.equals(val);
+        }
+
+        /** {@inheritDoc} */
+        @Override
+        public Value convertTo(int targetType) {
+            if (getType() == targetType) {
+                return this;
+            }
+            switch (targetType) {
+                case Value.BYTES: {
+                    return ValueBytes.getNoCopy(JdbcUtils.serialize(val, null));
+                }
+                case Value.STRING: {
+                    return ValueString.get(val.toString());
+                }
+                case Value.DOUBLE: {
+                    assert val.im == 0;
+                    return ValueDouble.get(val.re);
+                }
+                case Value.JAVA_OBJECT: {
+                    return ValueJavaObject.getNoCopy(JdbcUtils.serialize(val, null));
+                }
+            }
+            
+            throw DbException.get(
+                    ErrorCode.DATA_CONVERSION_ERROR_1, getString());
+        }
+
+        /** {@inheritDoc} */
+        @Override
+        public Value add(Value value) {
+            ValueComplex v = (ValueComplex)value;
```

Comment:
I modelled it after the ValueInt.add() and the latter doesn't attempt any conversions. So I think it is good enough.